File: src/main/java/core/procedures/vectors/VectorSet.java
package core.procedures.vectors;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.Type;
import core.scm.MutableVector;
import core.scm.Void;

public final class VectorSet extends AFn {

  public VectorSet() {
    super(new FnArgsBuilder().min(3).max(3)
                             .mandatory(new Class[]{MutableVector.class, Type.ExactNonNegativeInteger.class}).build());
  }

  @Override
  public String getName() {
    return "vector-set!";
  }

  @Override
  public Object apply3(Object arg1, Object arg2, Object arg3) {
    MutableVector vec = (MutableVector)arg1;
    Long pos = ((Number)arg2).longValue();
    if (pos >= vec.length()) {
      throw new IndexOutOfBoundsException(String.format("%s: value out of range: %s", getName(), pos));
    }
    vec.set(pos.intValue(), arg3);
    return Void.VOID;
  }
}

File: src/main/java/core/procedures/vectors/VectorSet.kt
package core.procedures.vectors

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.Type
import core.scm.MutableVector

class VectorSet : AFn<Any?, Unit>(name = "vector-set!", arity = Exactly(3),
                      mandatoryArgsTypes = arrayOf(MutableVector::class.java, Type.ExactNonNegativeInteger::class.java)) {

    override operator fun invoke(arg1: Any?, arg2: Any?, arg3: Any?) = (arg1!! as MutableVector).set((arg2!! as Number).toInt(), arg3)
}
